### ISSUE TITLE
sql: `ALTER TYPE DROP VALUE` and `CREATE FUNCTION` can misbehave within the same transaction

### ISSUE THREAD

**Issue by chrisseto** [cockroachdb/cockroach#114844]
**Describe the problem**

Run `cockroach demo` and execute the following SQL:

```sql
CREATE TYPE foo AS ENUM ('1', '2', '3');
BEGIN;
ALTER TYPE foo DROP VALUE '3';
CREATE FUNCTION udf_123(IN x foo) RETURNS VOID LANGUAGE SQL AS $$ SELECT NULL $$;
COMMIT;
```

Rather than succeeding, the following error is returned:

```
ERROR: transaction committed but schema change aborted with error: (42P01): could not validate enum value removal for "3": relation "[114]" does not exist
SQLSTATE: XXA00
HINT: Some of the non-DDL statements may have committed successfully, but some of the DDL statement(s) failed.
Manual inspection may be required to determine the actual state of the database.
```

**Expected behavior**

I expect that the transaction to successfully commit or for a more understandable error to be returned.

I suspect that this issue isn't unique to `ALTER TYPE DROP VALUE` and `CREATE FUNCTION`. It seems that the issue stems from `CREATE FUNCTION` update the referencing IDs field causes the validation on the type descriptor to fail.

**Context**

might relate to https://github.com/cockroachdb/cockroach/issues/113859

Jira issue: CRDB-33710

Epic CRDB-31325

**Comment by rimadeodhar** [comment]
This will likely be fixed by https://github.com/cockroachdb/cockroach/pull/118780. I'll assign this to myself to unblock once the PR is merged.

**Comment by rafiss** [comment]
fixed by #118780
